hail/python/hail/nd/nd.py

```diff
@@ -314,3 +314,84 @@ def concatenate(nds, axis=0):
     concat_ir = NDArrayConcat(makearr._ir, axis)
 
     return construct_expr(concat_ir, concat_ir.typ)
+
+
+@typecheck(N=expr_numeric, M=nullable(expr_numeric), dtype=HailType)
+def eye(N, M=None, dtype=hl.tfloat64):
+    """
+    Construct a 2-D :class:`.NDArrayExpression` with ones on the *main* diagonal
+    and zeros elsewhere.
+
+    Parameters
+    ----------
+    N : :class:`.NumericExpression` or Python number
+      Number of rows in the output.
+    M : :class:`.NumericExpression` or Python number, optional
+      Number of columns in the output. If None, defaults to `N`.
+    dtype : numeric :class:`.HailType`, optional
+      Element type of the returned array. Defaults to :py:data:`.tfloat64`
+
+    Returns
+    -------
+    I : :class:`.NDArrayExpression` representing a Hail ndarray of shape (N,M)
+      An ndarray whose elements are equal to one on the main diagonal, zeroes elsewhere.
+
+    See Also
+    --------
+    :func:`.identity`
+    :func:`.diagonal`
+
+    Examples
+    --------
+    >>> hl.eval(hl.nd.eye(2, dtype=hl.tint32))
```

Comment:
I think this should have one example where you pass `M`, otherwise I think we are good to merge.